experimental/lib/Analysis/GsaAnalysis.cpp

```diff
@@ -0,0 +1,485 @@
+    auto foundGate = std::find_if(list.begin(), list.end(),
+                                  [&](auto &t) {
+                                    return t->getArgumentNumber() ==
+                                           missing.ba.getArgNumber();
+                                  }
+
+    );
+    assert(foundGate != list.end() && "[GSA] Not found phi to reconnect");
+    missing.pi->input = *foundGate;
+  }
+
+  // Convert phis to MUs and GAMMAs
+  convertPhiToMu(funcOp);
+  convertPhiToGamma(funcOp);
+  printGateList();
+}
+
+template <typename FunctionType>
+void GsaAnalysis<FunctionType>::convertPhiToGamma(FunctionType &funcOp) {
+
+  if (funcOp.getBlocks().size() == 1)
+    return;
+
+  mlir::DominanceInfo domInfo;
+  mlir::CFGLoopInfo loopInfo(domInfo.getDomTree(&funcOp.getBody()));
+
+  // For each block
+  for (auto const &[phiBlock, phis] : gateList) {
+
+    // For each phi
+    for (auto *phi : phis) {
+
+      // Skip if the phi is not of type `Phi`
+      if (phi->gsaGateFunction != PhiGate)
+        continue;
+
+      // Get the phi operands and sort them according to the index of the basic
+      // block, so that input `i` comes before input `j` if
+      // `index_Bi`<`index_Bj`
+      auto phiOperands = phi->operands;
+
+      llvm::sort(phiOperands.begin(), phiOperands.end(),
+                 [](GateInput *a, GateInput *b) {
+                   return lessThanBlocks(a->getBlock(), b->getBlock());
+                 });
+
+      // Find the nearest common dominator among all the blocks involved
+      // in the phi inputs. Also add all the blocks to the list of 'blocks to
+      // avoid'
+      std::vector<Block *> blocksToAvoid = {phiOperands[0]->getBlock()};
+      Block *commonDominator = phiOperands[0]->getBlock();
+      for (auto &operand : llvm::drop_begin(phiOperands)) {
+        Block *bb = operand->getBlock();
+        commonDominator =
+            domInfo.findNearestCommonDominator(bb, commonDominator);
+        blocksToAvoid.push_back(bb);
+      }
+
+      // List of conditions present in all the expressions for all the paths
+      std::vector<std::string> conditionsList;
+
+      // Vector associating each input of the Phi to a boolean expression
+      std::vector<std::pair<boolean::BoolExpression *, GateInput *>>
+          expressionsList;
+
+      // For each input of the phi, compute the boolean expression which defines
+      // its usage
+      for (auto &operand : phiOperands) {
+
+        // Remove the current operand from the list of blocks to avoid
+        blocksToAvoid.erase(blocksToAvoid.begin());
+
+        // Find all the paths from `commonDominator` to `phiBlock` which pass
+        // through operand's block but not through any of the `blocksToAvoid`
+        auto paths = findAllPaths(commonDominator, phiBlock,
+                                  operand->getBlock(), blocksToAvoid);
+
+        boolean::BoolExpression *phiInputCondition =
+            boolean::BoolExpression::boolZero();
```

Comment:
You're repeating `boolean::` a lot, just import the namespace to avoid the redundancy.